### ISSUE TITLE
ZNE to UVW rotation feature

### ISSUE THREAD

**Issue by aringler-usgs** [obspy/obspy#1992]
Would obspy have interest in having a feature that rotates data from ZNE to UVW (Galperin orientation) as well as UVW back to ZNE?  

I know this is used a fair bit by network operators to identify potentially noisy components or when a physical module has become problematic.  It is also occasionally used when a sensor has gone into UVW mode and someone would still like to make use of the data in traditional ZNE coordinates.

If this would be valuable, then @amkearns-usgs can open a pull request to add this feature.  We understand if this might not be valuable to the general data user. 

**GitHub should mainly be used for bug reports and code developments/discussions. For generic questions, please use either the [[obspy-users] mailing list](http://lists.swapbytes.de/mailman/listinfo/obspy-users) (for scientific questions to the general ObsPy community) or our [gitter chat](https://gitter.im/obspy/obspy) (for technical questions and direct contact to the developers).**

---

Before submitting an Issue, please review the [Issue Guidelines](https://github.com/obspy/obspy/blob/master/CONTRIBUTING.md#submitting-an-issue).

* Please check whether the bug was already reported or fixed.
* Please provide the following information:
  -  ObsPy version, Python version and Platform (Windows, OSX, Linux ...)
  -  How did you install ObsPy and Python (pip, anaconda, from source ...)
  -  If possible please supply a short, self contained, correct example that
     demonstrates the issue.
  -  If this is a regression (used to work in an earlier version of ObsPy),
     please note when it used to work.

**Comment by krischer** [comment]
ObsPy can already do it: https://docs.obspy.org/packages/autogen/obspy.signal.rotate.rotate2zne.html

It goes from any three not linearly dependent components to ZNE. There is an `inverse` argument which can be used to reverse the whole procedure to go for example from ZNE to Galperin. There are a lot of test in this [file](https://github.com/obspy/obspy/blob/master/obspy/signal/tests/test_rotate.py) (also a few related to the Galperin configuration) as we struggled for a while to get it correct. But we are confident it does the right calculations now.

In general we are very interested in operator specific things. The "Obs" in ObsPy stands for observatory after all.

**Comment by megies** [comment]
See also https://github.com/obspy/obspy/issues/1913#issuecomment-336155562 for some code snippets around rotating to/from UVW..

**Comment by megies** [comment]
@LMurrayBergquist maybe you could work on this one. Basically that would be adding an option `->UVW` to `Stream.rotate()`. Easiest probably would be to internally call a rotate to `->ZNW` first and then use `obspy.signal.rotate.rotate2zne` as described above and in the linked comments with the appropriate angles and dips for Galperin orientation. If it's unclear you could also get in touch with @DominikStr.

**Comment by DominikStr** [comment]
@LMurrayBergquist are you already working on this one? If not I would give it a try.

**Comment by LMurrayBergquist** [comment]
> @LMurrayBergquist are you already working on this one? If not I would give it a try.

Yep, I started working on this one, but I'm not that far in, trying to figure out the right angles and how to get the components to put in signal.rotate.rotate2zne, so if you've got that you could take it!

**Comment by LMurrayBergquist** [comment]
@megies I could not find the station XX.TRC1 that you used in the code in the comment that you linked here, it seems like it's not available in any of the fdsn clients, do you know the location? I can build an inventory to use in this test case, but I need the latitude, longitude, and elevation of TRC1

**Comment by pantje** [comment]
When searching for a convenient way (single-function call) to rotate seismometer components in ObsPy to UVW (Galperin arrangement), I found this thread and a few hints to efforts made to implement this.

In all these I do not find that the type of sensor would be distinguished. In fact the definition of UVW with respect to XYZ is different for Streckseien STS-2 seismometers and Nanometrics seismometers. A general function would not work for both. How is this handled in ObsPy? Is it handled at all?

**Comment by filefolder** [comment]
revisiting this re: https://github.com/obspy/obspy/pull/2397 (not currently handled in ObsPy but would be nice for 1.5)

I don't think anyone can rely on channel metadata to define sensor type in any predictable way, so would have to implement two distinct UVW transforms that can be selected and mention common makes for each in the `rotate` text.

            ``'->UVW'``: Rotates data from three components into UVW components
                (Nanometrics-style Galperin orientation).
            ``'->UVWb'``: Rotates data from three components into UVW components
                (Streckheisen-style Galperin orientation).


I am not sure this is correct or which is which but
```
        elif method == "->UVW":
            from obspy.signal.rotate import rotate2zne
            # First rotate to ZNE as above, then rotate zne to uvw:
            zne_stream = self._rotate_to_zne(inventory, **kwargs)
            z = zne_stream[0]
            n = zne_stream[1]
            e = zne_stream[2]
            return rotate2zne(z, 90, -35.3, n, 330, -35.3, e, 210, -35.3,
                            inverse=True)
        # e.g. https://streckeisen.swiss/assets/downloads/sts-2-manual.pdf
        elif method == "->UVWb":
            from obspy.signal.rotate import rotate2zne
            # First rotate to ZNE as above, then rotate zne to uvw:
            zne_stream = self._rotate_to_zne(inventory, **kwargs)
            z = zne_stream[0]
            n = zne_stream[1]
            e = zne_stream[2]
            return rotate2zne(? not sure
                            inverse=True)
```
